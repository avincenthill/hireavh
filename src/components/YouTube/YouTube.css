.dropdown-label {
  margin-right: 2vmin;
}

.video-list {
  margin-top: 3vmin;
}

.video {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 15vmin;
  text-decoration: none;
  color: var(--c-black);
  margin-bottom: 3vh;
  border-radius: var(--s-xs);
  border: 3px var(--c-1) solid;
  background-color: var(--c-2);
}

.video-error {
  margin-bottom: 5vmin;
}

.video-img {
  height: 100%;
  border-right: 3px var(--c-1) solid;
  border-radius: var(--s-xs) 0 0 var(--s-xs);
}

.video-metadata {
  text-align: left;
  margin-left: 2vmin;
  width: 60vw;
}

.video-metadata-title {
  padding: 0;
  margin: 0;
  font-size: var(--fs-xs);
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  width: 59vw;
}

.video-metadata-description {
  display: none;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  width: 59vw;
}

.video-metadata-channel-and-date {
  font-size: var(--fs-xxs);
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  width: 59vw;
}

.channels-list {
  padding: 0;
  margin-bottom: 5vmin;
  list-style-type: none;
}

.channels-list-item {
  margin: 0.5vmin;
}

.channels-list-item-link {
  color: var(--c-white);
}

/* large devices */
@media (min-width: 775px) {
  .video {
    height: 10vmin;
  }
  .video-metadata-title {
    padding: 0;
    margin: 0;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    font-size: var(--fs-m);
  }
  .video-metadata-description {
    display: block;
    font-size: var(--fs-xs);
  }
  .video-metadata-channel-and-date {
    display: block;
    font-size: var(--fs-xs);
  }
}
