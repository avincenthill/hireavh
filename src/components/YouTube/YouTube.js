import './YouTube.css';
import { NavList } from 'components/Nav/Nav';
import Page from 'components/Page/Page';
import React from 'react';
import channelsByType from './data/channelsByType';
// TBD: refactor out static content to content link
// import content from 'content/content';
import mockVideoData from './data/mockVideoData';

class YouTube extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      YOUTUBE_API_KEY: 'AIzaSyBCdXhhkF8p1w-SDM0GFiia2Huhz5WQyR8',
      selectedType: 'Mock (testing)',
      currentVideos: [],
      videosByType: {
        'Mock (testing)': mockVideoData.items,
      },
      channelsByType,
    };
  }

  componentDidMount = () => {
    this.fetchVideos();
  };

  renderChannels = () => {
    if (
      !this.state.channelsByType[this.state.selectedType] ||
      !this.state.channelsByType[this.state.selectedType].length
    ) {
      return <li className="channels-list-item">NO CHANNELS FOUND</li>;
    }
    return this.state.channelsByType[this.state.selectedType].map((channel) => {
      return (
        <li key={channel.id} className="channels-list-item">
          <a
            className="channels-list-item-link"
            href={`https://www.youtube.com/channel/${channel.id}/videos?view=0&sort=p/`}
          >
            {channel.name}
          </a>
        </li>
      );
    });
  };

  renderDropdownOptions = () => {
    return Object.keys(this.state.channelsByType).map((type) => {
      return (
        <option key={type} value={type}>
          {type}
        </option>
      );
    });
  };

  handleSelect = (event) => {
    this.setState(
      { selectedType: event.target.value, currentVideos: [] },
      () => {
        this.fetchVideos();
      }
    );
  };

  fetchVideos = () => {
    // no channels
    if (!this.state.channelsByType[this.state.selectedType]) {
      return;
    }

    // videos are cached in state
    if (
      this.state.videosByType[this.state.selectedType] &&
      this.state.videosByType[this.state.selectedType].length
    ) {
      this.setState({
        currentVideos: this.state.videosByType[this.state.selectedType],
      });
      return;
    }

    const channels = this.state.channelsByType[this.state.selectedType];
    const requestOptions = {
      method: 'GET',
      redirect: 'follow',
    };
    Promise.all(
      channels.map((channel) =>
        fetch(
          `https://www.googleapis.com/youtube/v3/search?key=${this.state.YOUTUBE_API_KEY}&channelId=${channel.id}&part=snippet,id&order=date&maxResults=5`,
          requestOptions
        )
      )
    )
      .then((responses) => Promise.all(responses.map((res) => res.text())))
      .then((texts) => {
        const fetchedVideos = texts
          .map((text) => {
            return JSON.parse(text).items;
          })
          .flat()
          .filter((video) => {
            if (video) return video;
          });
        const videosByType = this.state.videosByType;
        videosByType[this.state.selectedType] = fetchedVideos;
        this.setState({
          currentVideos: fetchedVideos,
          videosByType,
        });
      })
      // TBD: display num searches left in day, handle errors, display "out of quota rejections", increase quota
      .catch((error) => {
        this.setState({ currentVideos: [] });
      });
  };

  renderVideos = () => {
    if (!this.state.currentVideos.length) {
      return (
        <div className="video-error">{`NO VIDEOS FOUND OR API QUOTA EXCEEDED`}</div>
      );
    }
    return this.state.currentVideos
      .sort(
        (a, b) =>
          new Date(b.snippet.publishTime).valueOf() -
          new Date(a.snippet.publishTime).valueOf()
      )
      .map((video) => {
        if (!video) return;
        return (
          <a
            key={video.id.videoId + video.snippet.title}
            className="video button-hover-light"
            href={`http://www.youtube.com/watch?v=${video.id.videoId}`}
            target="_blank"
          >
            <img
              className="video-img"
              src={video.snippet.thumbnails.high.url}
              alt="video thumbnail img"
            />
            <div className="video-metadata">
              <h3 className="video-metadata-title">
                {video.snippet.title
                  .replace(/&quot;/g, '"')
                  .replace(/&#39;/g, "'")}
              </h3>
              <div className="video-metadata-description">
                {video.snippet.description}
              </div>
              <div className="video-metadata-channel-and-date">
                <i>{`${new Date(
                  video.snippet.publishTime
                ).toDateString()} - ${video.snippet.channelTitle}`}</i>
              </div>
            </div>
          </a>
        );
      });
  };

  render() {
    return (
      <div>
        <Page>
          <div>
            <h1>recent videos by select channels</h1>
            <br></br>
            <label className="dropdown-label">choose a content category:</label>
            <select
              name="type"
              id="type"
              onChange={this.handleSelect}
              value={this.state.selectedType}
            >
              {this.renderDropdownOptions()}
            </select>
          </div>
          <br></br>
          <div className="video-list">{this.renderVideos()}</div>
          <ul className="channels-list">
            <label>featured channels:</label>
            {this.renderChannels()}
          </ul>
          <br></br>
          <NavList></NavList>
          <hr></hr>
        </Page>
      </div>
    );
  }
}

export default YouTube;
