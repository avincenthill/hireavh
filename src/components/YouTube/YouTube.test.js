import React from 'react';
import YouTube from './YouTube';
import renderer from 'react-test-renderer';

describe('YouTube', () => {
  describe('snapshot', () => {
    it('should match the last saved snapshot', () => {
      const tree = renderer.create(<YouTube />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
