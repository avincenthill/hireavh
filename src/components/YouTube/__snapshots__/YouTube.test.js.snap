// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YouTube snapshot should match the last saved snapshot 1`] = `
<div>
  <div>
    <div>
      <div
        className="nav nav-top"
      >
        <div
          className="nav-icon $
          this.props.isBottom ? \\"nav-icon-bottom\\" : \\"nav-icon-top\\"
        }"
        >
          <svg
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-1)",
                "height": "3rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 8 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 4v1H0V4h8zM0 8h8V7H0v1zm0 3h8v-1H0v1z"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <a
        className="header-container"
        href="/about"
      >
        <div
          className="header-switch-theme"
        >
          <div
            className="toggle-switch"
          >
            <input
              className="toggle-switch-checkbox"
              defaultChecked={true}
              id="toggle-switch-theme"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="toggle-switch-label"
              htmlFor="toggle-switch-theme"
            >
              <span
                className="toggle-switch-inner"
                data-no="üåô"
                data-yes="‚òÄÔ∏è"
              />
            </label>
          </div>
        </div>
        üè† about
      </a>
    </div>
    <div
      className="page-container"
    >
      <div>
        <h1>
          recent videos by select channels
        </h1>
        <br />
        <label
          className="dropdown-label"
        >
          choose a content category:
        </label>
        <select
          id="type"
          name="type"
          onChange={[Function]}
          value="Mock (testing)"
        >
          <option
            value="Art and Music"
          >
            Art and Music (3)
          </option>
          <option
            value="Business and Law"
          >
            Business and Law (3)
          </option>
          <option
            value="Chess"
          >
            Chess (8)
          </option>
          <option
            value="Commentary"
          >
            Commentary (6)
          </option>
          <option
            value="Computer Simulations and Game Development"
          >
            Computer Simulations and Game Development (7)
          </option>
          <option
            value="Cooking"
          >
            Cooking (6)
          </option>
          <option
            value="Film"
          >
            Film (11)
          </option>
          <option
            value="Gaming"
          >
            Gaming (2)
          </option>
          <option
            value="Hacking and Cybersecurity"
          >
            Hacking and Cybersecurity (6)
          </option>
          <option
            value="Infotainment"
          >
            Infotainment (7)
          </option>
          <option
            value="Mathematics"
          >
            Mathematics (7)
          </option>
          <option
            value="Mock (testing)"
          >
            Mock (testing) (1)
          </option>
          <option
            value="Programming and Computer Science"
          >
            Programming and Computer Science (12)
          </option>
          <option
            value="Science and Engineering"
          >
            Science and Engineering (18)
          </option>
          <option
            value="True Crime"
          >
            True Crime (2)
          </option>
        </select>
      </div>
      <br />
      <div
        className="video-list"
      >
        <a
          className="video button-hover-light"
          href="http://www.youtube.com/watch?v=FrG4TEcSuRg"
          target="_blank"
        >
          <img
            alt="video thumbnail img"
            className="video-img"
            src="https://i.ytimg.com/vi/FrG4TEcSuRg/hqdefault.jpg"
          />
          <div
            className="video-metadata"
          >
            <h3
              className="video-metadata-title"
            >
              PSY - DADDY(feat. CL of 2NE1) M/V
            </h3>
            <div
              className="video-metadata-description"
            >
              #PSY #DADDY #ÎÇòÌåîÎ∞îÏßÄ #NapalBaji #7thALBUM #7ÏßëÏã∏Ïù¥Îã§ More about PSY@ http://www.youtube.com/officialpsy http://www.facebook.com/officialpsy ...
            </div>
            <div
              className="video-metadata-channel-and-date"
            >
              <i>
                Mon Nov 30 2015 - officialpsy
              </i>
            </div>
          </div>
        </a>
        <a
          className="video button-hover-light"
          href="http://www.youtube.com/watch?v=HkMNOlYcpHg"
          target="_blank"
        >
          <img
            alt="video thumbnail img"
            className="video-img"
            src="https://i.ytimg.com/vi/HkMNOlYcpHg/hqdefault.jpg"
          />
          <div
            className="video-metadata"
          >
            <h3
              className="video-metadata-title"
            >
              PSY - HANGOVER (feat. Snoop Dogg) M/V
            </h3>
            <div
              className="video-metadata-description"
            >
              #PSY #HANGOVER More about PSY@ http://www.psypark.com/ http://www.youtube.com/officialpsy http://www.facebook.com/officialpsy ...
            </div>
            <div
              className="video-metadata-channel-and-date"
            >
              <i>
                Sun Jun 08 2014 - officialpsy
              </i>
            </div>
          </div>
        </a>
        <a
          className="video button-hover-light"
          href="http://www.youtube.com/watch?v=ASO_zypdnsQ"
          target="_blank"
        >
          <img
            alt="video thumbnail img"
            className="video-img"
            src="https://i.ytimg.com/vi/ASO_zypdnsQ/hqdefault.jpg"
          />
          <div
            className="video-metadata"
          >
            <h3
              className="video-metadata-title"
            >
              PSY - GENTLEMAN M/V
            </h3>
            <div
              className="video-metadata-description"
            >
              #PSY #Ïã∏Ïù¥ #GENTLEMAN #Ï††ÌãÄÎß® More about PSY@ http://www.youtube.com/officialpsy http://www.facebook.com/officialpsy http://twitter.com/psy_oppa ...
            </div>
            <div
              className="video-metadata-channel-and-date"
            >
              <i>
                Sat Apr 13 2013 - officialpsy
              </i>
            </div>
          </div>
        </a>
        <a
          className="video button-hover-light"
          href="http://www.youtube.com/watch?v=wcLNteez3c4"
          target="_blank"
        >
          <img
            alt="video thumbnail img"
            className="video-img"
            src="https://i.ytimg.com/vi/wcLNteez3c4/hqdefault.jpg"
          />
          <div
            className="video-metadata"
          >
            <h3
              className="video-metadata-title"
            >
              PSY (ft. HYUNA) - Ïò§Îπ§ Îî± ÎÇ¥ Ïä§ÌÉÄÏùº M/V
            </h3>
            <div
              className="video-metadata-description"
            >
              Psy's 2nd Version of "GANGNAM STYLE" by Hyuna Released! Gangnam Style - Hyuna Ver. was discussed during the making of the music video for "Gangnam ...
            </div>
            <div
              className="video-metadata-channel-and-date"
            >
              <i>
                Tue Aug 14 2012 - officialpsy
              </i>
            </div>
          </div>
        </a>
        <a
          className="video button-hover-light"
          href="http://www.youtube.com/watch?v=9bZkp7q19f0"
          target="_blank"
        >
          <img
            alt="video thumbnail img"
            className="video-img"
            src="https://i.ytimg.com/vi/9bZkp7q19f0/hqdefault.jpg"
          />
          <div
            className="video-metadata"
          >
            <h3
              className="video-metadata-title"
            >
              PSY - GANGNAM STYLE(Í∞ïÎÇ®Ïä§ÌÉÄÏùº) M/V
            </h3>
            <div
              className="video-metadata-description"
            >
              #PSY #Ïã∏Ïù¥ #GANGNAMSTYLE #Í∞ïÎÇ®Ïä§ÌÉÄÏùº More about PSY@ http://www.youtube.com/officialpsy http://www.facebook.com/officialpsy ...
            </div>
            <div
              className="video-metadata-channel-and-date"
            >
              <i>
                Sun Jul 15 2012 - officialpsy
              </i>
            </div>
          </div>
        </a>
      </div>
      <ul
        className="channels-list"
      >
        <label>
          featured channels:
        </label>
        <li
          className="channels-list-item"
        >
          <a
            className="channels-list-item-link"
            href="https://www.youtube.com/channel/UCrDkAvwZum-UTjHmzDI2iIw/videos?view=0&sort=p/"
            target="_blank"
          >
            officialpsy
          </a>
        </li>
      </ul>
      <br />
      <div
        className="nav-list"
      >
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/about"
            >
              <span>
                üè† about
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/blog"
            >
              <span>
                ‚úçÔ∏è blog
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/projects"
            >
              <span>
                üì¶ projects
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/resume"
            >
              <span>
                üìÑ resume
              </span>
            </a>
          </h2>
        </div>
        <div>
          <h2
            className="link"
          >
            <a
              className="link-a"
              href="/contact"
            >
              <span>
                üìá contact
              </span>
            </a>
          </h2>
        </div>
      </div>
      <hr />
    </div>
    <div>
      <div
        className="nav nav-bottom"
      >
        <div
          className="nav-icon $
          this.props.isBottom ? \\"nav-icon-bottom\\" : \\"nav-icon-top\\"
        }"
        >
          <svg
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-1)",
                "height": "3rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 8 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 4v1H0V4h8zM0 8h8V7H0v1zm0 3h8v-1H0v1z"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <a
        className="footer-container"
        href="/about"
      >
        <span>
          üè† about
        </span>
      </a>
    </div>
  </div>
</div>
`;
